bam_consensus.c

```diff
@@ -468,63 +958,69 @@ int nm_init(void *client_data, samFile *fp, sam_hdr_t *h, pileup_t *p) {
-            local_nm[i] = t < qual[i] ? t : qual[i];
-            if (qmin > qual[i+qhalo])
-                qmin = qual[i+qhalo];
-            else if (qmin <= qual[i-qhalo]) {
+            if (opts->homopoly_fix && bam_seqi(seq, i) != base) {
+                polyl = i;
+                base = bam_seqi(seq, i);
+                qminp = qual[i];
                 int j;
-                qmin = 50;
-                for (j = i-qhalo+1; j <= i+qhalo; j++)
-                    if (qmin > qual[j])
-                        qmin = qual[j];
+                for (j = i+1; j < qlen; j++) {
+                    if (bam_seqi(seq, j) != base)
+                        break;
+                    if (i < qhalop && qminp > qual[j])
+                        qminp = qual[j];
+                }
+                polyr = j-1;
+            } else {
+                // CHECK: do we want to have opts->homopoly_fix above,
+                // so when not in use we don't define pl to non-zero?
+                // Test on SynDip
+                polyr = polyl;
             }
-        }
-        for (; i < qlen; i++) {
-            local_nm[i] = qual[i];
-            local_nm[i] = (local_nm[i] + 6*qmin)/4;
-        }
+            int pl = polyr-polyl;
 
-        for (i = 0; i < qlen; i++) {
-            qual[i] = local_nm[i];
+            // Useful for SNPS, as we're judging the variation in
+            // length as an indicator for base-misalignment.
+            // Not so useful for indel calling where the longer the indel
+            // the less confident we are on the len variation being real.
+            int t = (opts->mode == MODE_BAYES_116)
+                ? (qual[i]   + 5*qmin)/4
+                : qual[i]/3 + (qminp-pl*2)*poly_adj;
 
-            // Plus overall rescale.
-            // Lower becomes lower, very high becomes a little higher.
-            // Helps deep GIAB, but detrimental elsewhere.  (What this really
-            // indicates is quality calibration differs per data set.)
-            // It's probably something best accounted for somewhere else.
 
-            //qual[i] = qual[i]*qual[i]/40+1;
-        }
-        memset(local_nm, 0, qlen * sizeof(*local_nm));
-#else
-        // Skew local NM by qual vs min-qual delta
-        uint8_t *qual = bam_get_qual(b);
-        const int qhalo = 8; // 4
-        int qmin = 99;
-        for (i = 0; i < qlen && i < qhalo; i++) {
-            if (qmin > qual[i])
-                qmin = qual[i];
-        }
-        for (;i < qlen-qhalo; i++) {
-            int t = (qual[i]   + 5*qmin)/4; // good on 15x
-            local_nm[i] += t < qual[i] ? (qual[i]-t) : 0;
+            local_nm[i] += t < qual[i] ? qual[i]-t : 0;
+
+            // Brute force qminp in polyl to polyr range.
+            // TODO: optimise this with sliding window
+            qminp = qual[i];
+            for (int j = MAX(polyl,i-qhalop); j <= MIN(polyr,i+qhalop); j++)
+                if (qminp < qual[j])
```

Comment:
This is finding the max rather than the min.  Is that correct?

Reply:
With `qminp` actually being the min value, it's working better mostly.

The first 140MB of chr1 for the HG002 PB Hifi sample:

![image](https://user-images.githubusercontent.com/2210525/216363068-692e6aed-4012-444d-9de2-3b964a42c2be.png)

![image](https://user-images.githubusercontent.com/2210525/216363467-32bb7f3c-00d7-4301-be87-ff9ed2d34417.png)

![image](https://user-images.githubusercontent.com/2210525/216363536-440a7183-802a-4dc7-a869-b90529c870e9.png)

A slight increase in FN for deletions, but a big drop in FP for insertions and a modest one for SNPs.  Overall a win.